torch/csrc/jit/graph_executor.cpp

```diff
@@ -409,70 +412,117 @@ struct GraphExecutorImpl {
+  // marked MaybeDefined and not optimized.
   void specializeUndef(Graph & g, const ArgumentSpec & spec) {
+    enum class State { Defined, Undefined, MaybeDefined };
+    std::unordered_map<Value*, State> state;
     for(size_t i = 0; i < spec.size(); i++) {
-      std::vector<Value*> to_replace;
-      // do not edit in place, since it invalidates uses iterator
-      for(auto u : g.inputs()[i]->uses()) {
-        if(u.user->kind() == prim::ReplaceIfUndef) {
-          to_replace.push_back(u.user->output());
-        }
-      }
-      for(auto v : to_replace) {
-        // if it is defined, then we replace with 'v' if not,
-        // we replace with 'replacement' which is normally just a zero tensor
-        int idx = spec.tensorInfo(i).defined() ? 0 : 1;
-        v->replaceAllUsesWith(v->node()->inputs()[idx]);
-        v->node()->destroy();
-      }
+      state[g.inputs()[i]] = spec.tensorInfo(i).defined() ? State::Defined : State::Undefined;
     }
-  }
-  // a + 0 -> a
-  // 0 + a -> a
-  void propagateZeros(Graph & g) {
     for(auto it = g.nodes().begin(); it != g.nodes().end(); ++it) {
-      if(it->kind() == aten::add && it->inputs().size() == 2 && at::Scalar(it->t(attr::alpha)).toDouble() == 1.0) {
-        if(isZero(it->inputs()[0])) {
-          it->output()->replaceAllUsesWith(it->inputs()[1]);
-          it.destroyCurrent();
-        } else if(isZero(it->inputs()[1])) {
-          it->output()->replaceAllUsesWith(it->inputs()[0]);
-          it.destroyCurrent();
-        }
+      auto n = *it;
+      switch(n->kind()) {
+        case prim::LinearBlock: {
+          auto all_undefined =
+              std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
+                return state[v] == State::Undefined;
+              });
+          // Property 1: if all the gradInputs to the LinearBlock are undefined
+          // then the gradOutputs are also zero and will be represented as undefined nodes
+          if(all_undefined) {
+            auto undef = g.createUndefined()->insertAfter(n)->output();
+            for(auto o : n->outputs()) {
+              o->replaceAllUsesWith(undef);
+            }
+          } else {
+          // Property 2: LinearBlocks are required to correctly handle combinations
+          // of defined and undefined inputs. They are expected to produce defined
+          // output tensors in this case.
+
+            // Remove the LinearBlock, splicing its body back into the surrounding block
+            auto body = n->blocks().at(0);
+            for(auto input : n->inputs()){
+              // we should never get into a situation when specializing a LinearBlock
+              // where we do not know if a value is defined since at the top level
+              // a gradient graph is composed of Linear nodes and AutogradAdds
+              // and LinearNodes only appear in these graphs
+              JIT_ASSERT(state[input] != State::MaybeDefined);
```

Comment:
How sure are you about this assert? :)

Reply:
It is true now, and I want anyone who makes it not true to consider the implications of breaking it. In particular, this pass will need to lower anything containing a MaybeDefined into an if statement. Currently autograd graphs only have GradOf and AutogradAdd nodes, neither of which can introduce a MaybeDefined. If another node is introduced into the gradient graph, we need to adjust this pass to preserve this invariant that is being checked, otherwise lots of conditionals will start to appear. I want that change to be visible when made hence the assert rather than handling the lowering case.